src/main/scala/fs2/aws/internal/Internal.scala

```diff
@@ -46,4 +49,38 @@ object Internal {
         implicit F: Effect[F]): F[ListenableFuture[UserRecordResult]] =
       F.delay(client.addUserRecord(streamName, partitionKey, data))
   }
+
+  /** Helper flow to group elements of a stream into K substreams.
+    * Grows with the number of distinct 'K' selectors
+    *
+    *  @tparam F effect type of the fs2 stream
+    *  @param selector partitioning function based on the element
+    *  @return a FS2 pipe producing a new sub-stream of elements grouped by the selector
+    */
+  def groupBy[F[_], A, K](selector: A => F[K])(
+      implicit F: Concurrent[F]): Pipe[F, A, (K, Stream[F, A])] = { in =>
+    Stream.eval(Ref.of[F, Map[K, Queue[F, Option[A]]]](Map.empty)).flatMap { st =>
+      val cleanup = {
+        import alleycats.std.all._
+        st.get.flatMap(_.traverse_(_.enqueue1(None)))
+      }
+
+      (in ++ Stream.eval_(cleanup))
+        .evalMap { el =>
+          (selector(el), st.get).mapN { (key, queues) =>
+            queues
+              .get(key)
+              .fold {
+                for {
+                  newQ <- Queue.unbounded[F, Option[A]] // Create a new queue
+                  _    <- st.modify(x => (x + (key -> newQ), x)) // Update the ref of queues
```

Comment:
Can we rename this magic `x` to `shardsToQueues` or something meaningful, it is hard to follow

Reply:
yes, i'll try to make it a bit more clear